discovery.go

```diff
@@ -0,0 +1,127 @@
+	"go.viam.com/rdk/logging"
+)
+
+// RTSPDiscovery is responsible for discovering RTSP camera devices using WS-Discovery and ONVIF.
+type RTSPDiscovery struct {
+	multicastAddress string
+	logger           logging.Logger
+}
+
+// NewRTSPDiscovery creates a new RTSPDiscovery instance with default values.
+func NewRTSPDiscovery(logger logging.Logger) *RTSPDiscovery {
+	return &RTSPDiscovery{
+		multicastAddress: "239.255.255.250:3702", // Standard WS-Discovery multicast address
+		logger:           logger,
+	}
+}
+
+// generateDiscoveryMessage formats an xml discovery message properly.
+func (d *RTSPDiscovery) generateDiscoveryMessage() string {
+	messageID := uuid.New().String()
+	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
+	<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope" 
+                xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" 
+                xmlns:wsdd="http://schemas.xmlsoap.org/ws/2005/04/discovery">
+		<SOAP-ENV:Header>
+			<wsa:MessageID>uuid:%s</wsa:MessageID>
+			<wsa:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To>
+			<wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</wsa:Action>
+		</SOAP-ENV:Header>
+		<SOAP-ENV:Body>
+			<wsdd:Probe>
+				<wsdd:Types>dn:NetworkVideoTransmitter</wsdd:Types>
+			</wsdd:Probe>
+		</SOAP-ENV:Body>
+	</SOAP-ENV:Envelope>`, messageID)
+}
+
+// discoverRTSPAddresses performs a WS-Discovery and extracts RTSP addresses from the XAddrs field.
+func (d *RTSPDiscovery) discoverRTSPAddresses() ([]string, error) {
+	var discoveredAddresses []string
+
+	addr, err := net.ResolveUDPAddr("udp4", d.multicastAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve UDP address: %w", err)
+	}
+
+	conn, err := net.ListenUDP("udp4", nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create UDP socket: %w", err)
+	}
+	defer conn.Close()
+
+	_, err = conn.WriteToUDP([]byte(d.generateDiscoveryMessage()), addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send discovery message: %w", err)
+	}
+
+	err = conn.SetReadDeadline(time.Now().Add(10 * time.Second))
+	if err != nil {
+		return nil, fmt.Errorf("failed to set read deadline: %w", err)
+	}
+
+	buffer := make([]byte, 1536)
+	for {
+		n, _, err := conn.ReadFromUDP(buffer)
+		if err != nil {
+			var netErr net.Error
+			if errors.As(err, &netErr) && netErr.Timeout() {
+				d.logger.Debug("Timed out after waiting.")
+				return discoveredAddresses, nil
+			}
+			return nil, fmt.Errorf("error reading from UDP: %w", err)
+		}
+
+		response := buffer[:n]
+		xaddrs, err := d.extractXAddrsFromProbeMatch(response)
+		if err != nil {
+			d.logger.Warnf("Failed to parse response: %w\n", err)
+			continue
+		}
+
+		discoveredAddresses = append(discoveredAddresses, xaddrs...)
+	}
+}
+
+// extractXAddrsFromProbeMatch extracts XAddrs from the WS-Discovery ProbeMatch response.
+func (d *RTSPDiscovery) extractXAddrsFromProbeMatch(response []byte) ([]string, error) {
```

Comment:
What is the goal of such a test? Just to unit test its parsing functionality ya?

Reply:
yes, we should also test the happy path, but I wonder if there are silent fail cases in your error checking after decoding, and we should  be defensive about cases such as those.

Reply:
Added some tests let me know what you think

Reply:
Okay so from my current understanding of the user flow - Discov is called when a resource is added to the graph in runtime. We error if we parse bad data but only warn the user once, so that's okay and not spammy; invalid, but nicely formed bytes do not error but do not return an address that is unusable either. Good to know.

We may want to just completely ignore malformed bytes, idk. Up to you there, because this API is servicing giving options to a user to select from, ultimately.

Reply:
Makes sense. Changed error return to a warning log